lib/github_changelog_generator/project_name_finder.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+module GitHubChangelogGenerator
+  class ProjectNameFinder
+    attr_reader :options, :args
+
+    # @param options [Options]
+    # @param args [Array] ARGV or the empty list.
+    #                     The first element of this argument takes two forms:
+    #                     Either a full GitHub URL, or a 'username/projectname',
+    #                     or simply a GitHub username.
+    #                     If the first element is given as a username,
+    #                     then the second element can be given as a projectname.
+    def initialize(options, args)
+      @options = options
+      @args = args
+    end
+
+    FIXED_TEST_PROJECT = %w(skywinder changelog_test)
+
+    NO_MATCHING_USER_AND_PROJECT = [nil, nil]
+
+    # Returns a tuple of user and project from CLI arguments or git remote.
+    #
+    # @return [Array<String>] user and project, or nil if unsuccessful
+    def call
+      [
+        -> { from_cli_option },
+        -> { FIXED_TEST_PROJECT if in_development? },
+        -> { from_git_remote }
+      ].find(proc { NO_MATCHING_USER_AND_PROJECT }) do |strategy|
+        user, project = strategy.call
+        break [user, project] if user && project
+      end
+    end
+
+    def in_development?
+      ENV["RUBYLIB"] =~ /ruby-debug-ide/
+    end
+
+    # Returns GitHub username and project from CLI arguments
+    #
+    # @return [Array, nil] user and project, or nil if unsuccessful
+    def from_cli_option
+      if args[0] && !args[1]
+        github_site_pattern =~ args.first
+
+        begin
+          param = Regexp.last_match(2).nil?
+        rescue
+          puts "Can't detect user and name from first parameter: '#{args.first}' -> exit'"
+          return
+        end
+        if param
+          return NO_MATCHING_USER_AND_PROJECT
+        else
+          [Regexp.last_match(:user), Regexp.last_match(:project)]
+        end
+      end
+    end
```

Comment:
Forgive me if I've overlooked something, since I'm only looking at this diff right now - but is this method and the following ones only public in order to support being called from the spec? In that case, it might be worth some effort trying to make them private I think.